pep-0692.rst

```diff
@@ -0,0 +1,438 @@
+    
+    def src(**kwargs: **Example): ...
+    def dest(animal: Dog, string: str, number: int = ...): ...
+
+    dest = src  # OK!
+
+3. The destination callable contains ``**kwargs: **TypedDict``, the source
+callable doesn't contain ``**kwargs`` and the destination function's
+``TypedDict`` fields are assignable to the corresponding keyword arguments in
+the source function. Again, required and not required keys should be compatible
+with the keyword arguments and the rest of the parameters should be compatible.
+For example:
+
+::
+
+    class Example(TypedDict):
+        animal: Dog
+        string: str
+        number: NotRequired[int]
+
+    def src(animal: Animal, string: str, number: int = ...): ...
+    def dest(**kwargs: **Example): ...
+
+    dest = src  # OK!
+
+4. The destination callable contains ``**kwargs: **TypedDict`` and the source
+callable contains untyped ``**kwargs``:
+
+::
+
+    def src(**kwargs): ...
+    def dest(**kwargs: **Movie): ...
+
+    dest = src  # OK!
+
+5. The destination callable contains ``**kwargs: **TypedDict``, the source
+callable contains traditionally typed ``**kwargs: T`` and each of the
+destination function ``TypedDict``\'s fields is assignable to a variable of type
+``T``.
+
+On the other hand, if the destination callable contains either untyped or
+traditionally typed ``**kwargs: T`` and the source callable is typed using
+``**kwargs: **TypedDict`` then an error should be generated, because
+traditionally typed ``**kwargs`` aren't checked for keyword names.
+
+To sum up points, functions' parameters should behave contravariantly. In
+addition, functions' return types should behave covariantly. 
+
+Grammar Changes
+===============
+This PEP requires a grammar change so that the double asterisk syntax is allowed
+for ``**kwargs`` annotations. The proposed change is to extend the ``kwds`` rule
+as follows:
+
+Before:
+
+::
+
+    kwds: '**' param_no_default 
+
+After:
+
+::
+
+    kwds:
+        | '**' param_no_default_double_star_annotation
+        | '**' param_no_default
+
+    param_no_default_double_star_annotation:
+        | param_double_star_annotation & ')'
+
+    param_double_star_annotation: NAME double_star_annotation
+
+    double_star_annotation: ':' double_star_expression
+
+    double_star_expression: '**' expression
+
+A new AST node needs to be created so that type checkers can differentiate the
+semantics of the new syntax and take into account that the ``**kwargs`` should
+be unpacked. The proposition is to add a new ``DoubleStarred`` AST node. Then,
+an AST node for the function defined as:
+
+::
+
+    def foo(**kwargs: **Movie): ...
+
+should look as below:
+
+::
+
+    FunctionDef(
+      name='foo',
+      args=arguments(
+        posonlyargs=[],
+        args=[],
+        kwonlyargs=[],
+        kw_defaults=[],
+        kwarg=arg(
+          arg='kwargs',
+          annotation=DoubleStarred(
+            value=Name(id='Movie', ctx=Load()),
+            ctx=Load())),
```

Comment:
So a `DoubleStarred(...)` would appear here if you wrote `def foo(**kwargs: **Movie): ...`.

What would appear instead if you wrote `def foo(**kwargs: Unpack[Movie]): ...`? (This may be a useful example to distinguish how both forms would appear in the AST, if they were to appear differently.)

Reply:
Added the AST with Unpack in the Backwards Compatibility section.